### PR TITLE
Add self-supervised 3D-Var-based AI data assimilation 

### PR DESCRIPTION
**Pull request by SOHAMPAL23**
# Pull Request

- Add self-supervised AI-based data assimilation using 3D-Var loss

## Description

- This PR introduces a self-supervised AI-based data assimilation prototype

- The implementation replaces supervised learning with a physics-based 3D-Var cost function, allowing the model to learn directly from:

- Sparse/noisy observations

## Scope

- Focused on demonstrating the feasibility and correctness of the self-supervised assimilation approach

- Intended as an experimental framework, not a full operational replacement

## Checklist

-  Implements the 3D-Var objective as the training loss

- Supports both forecast-based and cold-start first-guess states

- Modular design for experimenting with different observation operators and error assumptions

- Suitable as a foundation for extending to real-world variables

### CODE REVIEW

File: graph_weather/models/aurora/model.py [hunk 1]
```diff
@@ -99,7 +147,18 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
 
 
 class SelfAttentionLayer(nn.Module):
+    """Self-attention layer for processing point cloud features.
```
Comment (jacobbieker) [inline graph_weather/models/aurora/model.py:150]: Please split these just doc changes into a separate PR, this is a huge PR that is hard to review and has a lot of unrelated changes. Do that, and then I can review just the implementation.

File: graph_weather/models/data_assimilation.py [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+
+    # Create H matrix (num_obs x total_size)
+    H = torch.zeros(num_obs, total_size)
+    for i, idx in enumerate(obs_indices):
+        H[i, idx] = 1.0
+
+    return H
+
+
+def generate_synthetic_data(batch_size=32, grid_size=(10, 10), num_channels=1):
```
Comment (jacobbieker) [inline graph_weather/models/data_assimilation.py]: This generation for testing shouldn't be in the model file, only in the unit tests

File: graph_weather/models/training_loop.py
Comment (jacobbieker) [inline graph_weather/models/training_loop.py:1]: These model training loops, etc. stuff shouldn't be in the `/models/` directory, but in a subdirectory, like `/models/data_assimilation/` or something similar

File: graph_weather/models/visualization.py
Comment (jacobbieker) [inline graph_weather/models/visualization.py:1]: These feel like more generic things for training that shouldn't be in a model PR. If you want to add more visualizations, feel free, but that should be separate.

File: graph_weather/data_assimilation_implementation.md
Comment (jacobbieker) [inline graph_weather/data_assimilation_implementation.md:1]: This feels very AI-generated, please remove. A short readme linking to the paper can be fine, look at the GenCast implementation here as an example, but it should be in the model's subfolder and be concise.

File: graph_weather/example_usage.py
Comment (jacobbieker) [inline graph_weather/example_usage.py:1]: This also seems unnecessary, just have a short example in the readme on how to load and run inference on the model, this isn't needed.

File: graph_weather/test_data_assimilation.py
Comment (jacobbieker) [inline graph_weather/test_data_assimilation.py:1]: Remove print statements, unit tests shouldn't have that, this should be under the tests subfolder, and shouldn't have a name == main block, as it needs to work with pytest.

File: graph_weather/models/evaluation.py
Comment (jacobbieker) [inline graph_weather/models/evaluation.py:1]: I don't think this evaluation should be included in this PR, a lot of the computation already exists somewhere in the repo, and is more generally useful. Please refactor or remove.

File: graph_weather/models/data_assimilation.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+        # Concatenate background and observations along the feature dimension
+        combined_input = torch.cat([background, observations], dim=-1)
+
+        # Pass through the network to get analysis
+        analysis = self.network(combined_input)
+
+        return analysis
+
+
+class SimpleDataAssimilationModel(nn.Module):
```
Comment (jacobbieker) [inline graph_weather/models/data_assimilation.py]: I would prefer just the model implementation from the paper, not a second, simpler one for this.

File: graph_weather/models/data_assimilation.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+            layers.append(nn.Linear(hidden_dim, hidden_dim))
+            layers.append(self.activation)
+            layers.append(nn.Dropout(dropout))
+
+        # Output layer to produce analysis
+        layers.append(nn.Linear(hidden_dim, input_dim))
+
+        self.network = nn.Sequential(*layers)
+
+    def forward(self, background, observations):
```
Comment (jacobbieker) [inline graph_weather/models/data_assimilation.py:192]: In this repo, we want the models to take in points and output points, so they are all compatible. Please make this work with the same kind of interface as the GraphWeatherForecaster, or GenCast implementations.

File: graph_weather/data/assimilation_dataloader.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+        sample = {"background": bg, "observations": obs}
+
+        if self.true_states is not None:
+            sample["true_state"] = self.true_states[idx]
+
+        return sample
+
+
+def create_synthetic_assimilation_dataset(
```
Comment (jacobbieker) [inline graph_weather/data/assimilation_dataloader.py]: Remove

File: graph_weather/data/assimilation_dataloader.py
Comment (jacobbieker) [inline graph_weather/data/assimilation_dataloader.py:1]: This might be needed, but needs changes. The models being added should be compatible with taking in points and lat/lon locations, and so should this one. This might make it not necessary to have this.

File: graph_weather/data/__init__.py [hunk 6]
```diff
@@ -1,5 +1,10 @@
 """Dataloaders and data processing utilities"""
 
-from .anemoi_dataloader import AnemoiDataset
+try:
```
Comment (jacobbieker) [inline graph_weather/data/__init__.py]: This should import fine, so remoev this change.

**Comment by SOHAMPAL23** [comment]
Sure @jacobbieker
 Working on the splitting part of the pr and making other changes in the new PR

**Comment by jacobbieker** [comment]
Closing this as #196 is a duplicate
